lockin/src/lib.rs

```diff
@@ -0,0 +1,973 @@
+    let mut total: f32 = 0.;
+    for val in x.iter() {
+        total += val;
+    }
+    total
+}
+
+/// Count number of valid TimeStamps from `tstamps`.
+fn tstamps_valid_count(tstamps: &[TimeStamp; 2]) -> usize {
+    let mut valid_count: usize = 0;
+    for i in 0..2 {
+        if tstamps[i].is_valid() {
+            valid_count += 1;
+        }
+    }
+    valid_count
+}
+
+/// Add new timestamps to the TimeStamp memory.
+///
+/// # Arguments
+///
+/// * `ref_tstamps` - New timestamp values.
+/// * `valid_tstamps` - Number of valid timestamps.
+/// * `tstamps_mem` - Last 2 recorded timestamps.
+fn record_new_tstamps(
+    ref_tstamps: [u16; TSTAMP_BUFFER_SIZE],
+    valid_tstamps: usize,
+    tstamps_mem: &mut [TimeStamp; 2],
+) {
+    if valid_tstamps > 1 {
+        tstamps_mem[1].new_count(ref_tstamps[valid_tstamps - 2]);
+        tstamps_mem[0].new_count(ref_tstamps[valid_tstamps - 1]);
+    } else if valid_tstamps == 1 {
+        tstamps_mem[1].count = tstamps_mem[0].count;
+        tstamps_mem[1].sequences_old = tstamps_mem[0].sequences_old;
+        tstamps_mem[0].new_count(ref_tstamps[valid_tstamps - 1]);
+    }
+}
+
+/// Map `iq_to_a` and `iq_to_t` to each pair of `i` and `q`.
+fn iq_to_at_map(
+    i: [f32; OUTPUT_BUFFER_SIZE],
+    q: [f32; OUTPUT_BUFFER_SIZE],
+) -> ([f32; OUTPUT_BUFFER_SIZE], [f32; OUTPUT_BUFFER_SIZE]) {
+    let mut a: [f32; OUTPUT_BUFFER_SIZE] = [0.; OUTPUT_BUFFER_SIZE];
+    let mut t: [f32; OUTPUT_BUFFER_SIZE] = [0.; OUTPUT_BUFFER_SIZE];
+    for k in 0..OUTPUT_BUFFER_SIZE {
+        a[k] = iq_to_a(i[k], q[k]);
+        t[k] = iq_to_t(i[k], q[k]);
+    }
+    (a, t)
+}
+
+/// Returns magnitude from in-phase and quadrature signals.
+///
+/// # Arguments
+///
+/// `i` - In-phase signal.
+/// `q` - Quadrature signal.
+fn iq_to_a(i: f32, q: f32) -> f32 {
+    2. * sqrt(pow2(i) + pow2(q))
+}
+
+/// Returns angle from in-phase and quadrature signals.
+///
+/// # Arguments
+///
+/// `i` - In-phase signal.
+/// `q` - Quadrature signal.
+fn iq_to_t(i: f32, q: f32) -> f32 {
+    atan2(q, i)
+}
+
+/// Demodulation phase values corresponding to each ADC sample.
+///
+/// # Arguments
+///
+/// * `first_t` - First timestamp value from the current processing
+/// period. The value provided here is ignored if there were no
+/// timestamps in the current processing period.
+/// * `tstamps_mem` - Recorded TimeStamps.
+/// * `phi` - Reference phase offset.
+/// * `fscale` - Frequency scaling factor for the demodulation signal.
+/// * `tadc` - ADC sampling period.
+///
+/// # Latency (ADC batch size = 16): 3.3us
+///
+/// 2.9us from computing `real_phase`.
+fn adc_phases(
+    first_t: u32,
+    tstamps_mem: &mut [TimeStamp; 2],
+    phi: u32,
+    fscale: u32,
+    tadc: u32,
+) -> [f32; SAMPLE_BUFFER_SIZE] {
+    let overflow_count: u32 = tadc * SAMPLE_BUFFER_SIZE as u32;
+    let tref_count: u32 = tstamps_diff(tstamps_mem, overflow_count);
+    let mut thetas: [f32; SAMPLE_BUFFER_SIZE] = [0.; SAMPLE_BUFFER_SIZE];
```

Comment:
Place variable declarations immediately where they're used so referencing what they are is simpler.